### ISSUE TITLE
[Bug]: ModalProvider loses parameter reference when closing popup

### ISSUE THREAD

**Issue by drma-tech** [Megabit/Blazorise#5288]
### Blazorise Version

all

### What Blazorise provider are you running on?

Bootstrap4

### Link to minimal reproduction, or a simple code snippet

https://github.com/drma-tech/LostReferencePopup/tree/master/LostReferencePopup

### Steps to reproduce

https://www.awesomescreenshot.com/video/24634047?key=6e7dea7043e561daaea2891e9c94e7fd

### What is expected?

the value/reference should only be changed intentionally

### What is actually happening?

When the popup is opened by the service (this does not happen with the embedded popup) the parameters that were loaded in this call are reset when the popup closes.

### What browsers are you seeing the problem on?

Chrome

**Comment by David-Moreira** [comment]
Indeed, currently the ModalProvider does not keep state very well. As when it needs to re-render modals it does not reuse the RenderFragment.

There's this related issue, where we want to make it so it's aware of the RenderMode that is set and manages state accordingly : https://github.com/Megabit/Blazorise/issues/4612

**Comment by drma-tech** [comment]
@David-Moreira  This has been open for exactly one year. Are you going to fix this now or leave it for later? because if it's not going to be fixed now, I need to find an alternative solution.

**Comment by drma-tech** [comment]
or if you have any workarounds, that works for me too.

**Comment by David-Moreira** [comment]
I'm not sure we should change the behaviour in middle of v1.4. Users might depend on how it behaves already. 

I'd do #4612 for **v1.5**  which should solve your problem in needing to keep state.

@stsrki thoughts?

---

Well in the meantime, for workaround... maybe you could keep the state in localStorage or something like that, and make sure that the `ModalService` parameters are loaded & saved in that centralized state? Would the be feasible?

Other than that, the usage of regular Modals. But this can be a little more cumbersome depending on your needs.

I'll update you if I think of any more ideas.

**Comment by stsrki** [comment]
I'd do it in 1.5. It's not so far until release.

**Comment by drma-tech** [comment]
ok. i will wait

**Comment by David-Moreira** [comment]
@drma-tech 
Refer to #5304

@stsrki can we revive nightly builds #5143 and have @drma-tech test the feature ahead of the release if he would be available to when we merge the feature in.

**Comment by stsrki** [comment]
> @stsrki can we revive nightly builds #5143 and have @drma-tech test the feature ahead of the release if he would be available to when we merge the feature in.

We can try to make it on MyGet.

**Comment by David-Moreira** [comment]
@drma-tech this issue should be fixed with the introduction of https://github.com/Megabit/Blazorise/pull/5304
The next release is scheduled for March.
I will go ahead and close this issue.

If you'd like to test the feature ahead of time and give us feedback, please use our dev packages for v1.5.0.
https://www.myget.org/gallery/blazorise

**Comment by drma-tech** [comment]
@David-Moreira 

I still haven't been able to understand what I have to change in the code to maintain the popup state. I also didn't see anything in the documentation (is it a different domain?). Do you have any code examples I can use? You can also change the project I submitted here.

**Comment by David-Moreira** [comment]
At the moment we do not have documentation live for dev / pre-releases.
Only way is to run the documentation locally or to look at the corresponding pages in the source code:

[Documentation/Blazorise.Docs/Pages/Docs/Services/ModalProvider/Examples/ModalProviderStatefulExample.razor](https://github.com/Megabit/Blazorise/blob/master/Documentation/Blazorise.Docs/Pages/Docs/Services/ModalProvider/Examples/ModalProviderStatefulExample.razor)
![image](https://github.com/Megabit/Blazorise/assets/22283161/f91fec9a-a475-426f-a208-31d1e69d85b5)

https://github.com/Megabit/Blazorise/blob/master/Documentation/Blazorise.Docs/Pages/Docs/Services/ModalProvider/ModalProviderPage.razor
![image](https://github.com/Megabit/Blazorise/assets/22283161/8925fee9-ed66-47a3-a437-dedce5b3b396)

**Comment by drma-tech** [comment]
@David-Moreira 

Well, in addition to not solving the problem, the close of the first popup stops working.

From what I understand, this stateful maintains the popup reference for future uses, but this has no relation to the problem I reported.

The problem is not opening the popup, but closing the popup. When I close the second popup, the reference to the first is lost, and enabling stateful doesn't change this at all. Maybe what can solve this is the ElementId, but from what I understand it is only used in Show() and not in Hide().

Is there something I don't understand or is that it? I uploaded updates to the project so you can check.

**Comment by David-Moreira** [comment]
Hello @drma-tech 
Thanks for taking the time to test the feature.
Maybe I miss understood your problem? I will take a look as soon as I have time.

**Comment by David-Moreira** [comment]
Hello @drma-tech 
You are right, you do not have the same problem that we specifically fixed.

It seems like you are binding an object reference as the `Parameter` of the component that's rendered by the `ModalProvider` and then trying to change it by a new one.
Since the RenderFragment delegate already captured that one object reference you passed in, it will "rebind" it once it refreshes. 
Would the following be okay for you? Leave the object reference be the same, but mutate the values inside it:
`PageState.ActionNumber += (RefClass value) => { Ref!.Value = value.Value; StateHasChanged(); };`

This should work and be a decent workaround for now?

---

As for the issue itself, I do believe you have a point, but we're just using regular blazor RenderTreeBuilder api internally, 
![image](https://github.com/Megabit/Blazorise/assets/22283161/4fac91f3-c538-4c75-87a6-fabab56c91a2)
I'm honestly not sure how could this specifically be handled. If any of you @drma-tech , @stsrki has any suggestion, I'll take it. 
I'll retag this issue as Possible Bug & Investigate.

**Comment by drma-tech** [comment]
Why first popup is refreshing if I'm interacting only with the second one?

**Comment by David-Moreira** [comment]
> Is the component or the blazor that generates this refresh? If it's from the component, just avoid calling refresh, if I'm not interacting with the popup.

You would just be avoiding the issue if that's the case.
We pretty much just iterate over a collection of modals with the render fragment built from the blazor api render tree builder.

> If it's Blazor, you can keep these parameters in memory and pass them via callback or any other way.

well that's how it's supposed to work... The reference is captured. Blazor recommends using the approach of EventCallbsck to notify parameter changes back to the consumer of the component, instead of mutating the value directly. (here the consumer would be the page that called the first pop-up) 
There must be something to it. But I'm personally not sure. I would like to first make sure we understand what's going on fully and wouldn't like to create weird hacky workarounds if it's avoidable.

If you are that sure it's an easy fix, You are free to help us solve the issue by coming up with a PR if you'd like. We'd really appreciate it. :) Cheers.

**Comment by David-Moreira** [comment]
I was spending a little more time over this, I believe I found out the issue with both your approach and Blazorise.

In your approach if you want to make sure the reference is correctly updated, you will need to provide something like a 
`[Parameter] public EventCallback<RefClass?> RefChanged { get; set; }` and propagate the value back to the consumer (The one that Shows the PopupBase)

![image](https://github.com/Megabit/Blazorise/assets/22283161/1f38ffff-7155-46d1-b4a0-fb0d5c7a3973)
This one needs to be told that the reference should now point somewhere else.
Something like this:
![image](https://github.com/Megabit/Blazorise/assets/22283161/1a2ef279-ac7f-4ea3-bd23-1d844a9c8488)

---

Don't bother reading next section, if you don't care about our technical details, 
**bottom line is I believe we'll improve this binding for your needs** as long as you make sure the ref is correctly updated on your end.

---

**As for our side**, I believe we should capture & re-evaluate the `Action<ModalProviderParameterBuilder<TComponent>> parameters`. (This is the fluent based helper we provide you with)

Currently we use it once, the time you click `Show()` to extract the parameters, and then it is pretty much a Dictionary that sends the boxed parameters into the **Blazor** internals. The problem is that this Dictionary was the only thing captured in the `RenderFragment` so it would always hold that same reference.

I made some successful attempts with something along these lines
![image](https://github.com/Megabit/Blazorise/assets/22283161/40ff29fd-0012-4052-bc77-bf8d7ec3ea6a)
I don't think the `ModalProvider` will be negatively affected in any way with this change, but I'll need to do some more testing later.

@stsrki any concerns? I'd patch this in **v1.4** if I don't identify any issues.

**Comment by drma-tech** [comment]
I had tried this strategy of updating the original reference, but as it wasn't working, I imagined that you cut the binding and it was something internal to the popup, so I didn't know what to do to resolve it.

If you cut the binding, would it solve the problem? Like, what was passed to the popup will not be affected by anything external, unless it is intentional.

**Comment by David-Moreira** [comment]
> I had tried this strategy of updating the original reference, but as it wasn't working, I imagined that you cut the binding and it was something internal to the popup, so I didn't know what to do to resolve it.

To resolve it as a temporary workaround I see two options:
- You keep the state somewhere else. Like your scoped service for instance.
- You create a CopyFrom method of some kind, where instead of changing the reference, you just copy the values from the other ref.

> If you cut the binding, would it solve the problem? Like, what was passed to the popup will not be affected by anything external, unless it is intentional.

Can you rephrase this? I'm not sure what you mean. We are thinking in re-evaluating the parambuilder you pass on to the ModalService which should be closer to how Blazor works in components. i.e
`<MyComponent MyParameter=MyModelRef />`
In the `MyModelRef`, blazor will continuously bind to the `MyComponent` the latest `MyModelRef` value.

We expect that `ModalService.Show<MyComponent >(x=> x.Add("MyParameter", MyModelRef ))` will do something similar by re-evaluating the `x.Add("MyParameter", MyModelRef )` part when the **RenderFragment** delegate re-executes.

**Comment by drma-tech** [comment]
@David-Moreira 

I updated to dev2. I made the changes you suggested, but it still doesn't work.

Have you tested these changes? I have already updated the project.

**Comment by David-Moreira** [comment]
Hello. 
This is scheduled for the next patch for 1.4, since it was determined as a bug.

**Comment by David-Moreira** [comment]
Oh, but you are right in assuming the change would also be in master.

@stsrki have you merged in 1.4 patch to master yet, you usually do it on release day right?

**Comment by stsrki** [comment]
Yes, it's already merged into 1.4. But, I cam always unmerge it if you think it is worth it.

**Comment by David-Moreira** [comment]
Ah? I'm asking whether have you merged it to master yet. Since drma was trying to test it through our master dev release.
Or just give us an idea of the next 1.4 release patch.

**Comment by stsrki** [comment]
Ah, sorry, I misunderstand. I will merge it now and run the pipeline.

**Comment by stsrki** [comment]
It is now released as 1.5.0-dev-3.

**Comment by drma-tech** [comment]
ok, now its working. :)
